File: sdk/python/polar/v2026_04/services/customer_portal/members.py
from __future__ import annotations

import typing

from polar.base import (
    AsyncServiceBase,
    SyncServiceBase,
    parse_response_json,
    parse_response_none,
)
from polar.v2026_04.errors import (
    HTTPValidationError,
    MembersAddMember400Error,
    MembersAddMember401Error,
    MembersAddMember403Error,
    MembersListMembers401Error,
    MembersListMembers403Error,
    MembersRemoveMember400Error,
    MembersRemoveMember401Error,
    MembersRemoveMember403Error,
    MembersRemoveMember404Error,
    MembersUpdateMember400Error,
    MembersUpdateMember401Error,
    MembersUpdateMember403Error,
    MembersUpdateMember404Error,
)
from polar.v2026_04.inputs import (
    CustomerPortalMemberCreate,
    CustomerPortalMemberUpdate,
)
from polar.v2026_04.outputs import (
    CustomerPortalMember,
    ListResourceCustomerPortalMember,
)


class MembersSync(SyncServiceBase):
    def list_members(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> ListResourceCustomerPortalMember:
        """
        List all members of the customer's team.

        Only available to owners and billing managers of team customers.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Raises:
            MembersListMembers401Error: Authentication required
            MembersListMembers403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/members",
            path_params={},
            query_params={
                "page": page,
                "limit": limit,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            401: MembersListMembers401Error,
            403: MembersListMembers403Error,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, ListResourceCustomerPortalMember, method_errors
        )

    def iter_list_members(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> typing.Generator[CustomerPortalMember, None, None]:
        """
        List all members of the customer's team.

        Only available to owners and billing managers of team customers.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Returns:
            A generator that yields items of type CustomerPortalMember.

        Raises:
            MembersListMembers401Error: Authentication required
            MembersListMembers403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = self.list_members(
                page=page,
                limit=limit,
            )
            yield from response.items
            if page >= response.pagination.max_page:
                break
            page += 1

    def add_member(
        self,
        **kwargs: typing.Unpack[CustomerPortalMemberCreate],
    ) -> CustomerPortalMember:
        """
        Add a new member to the customer's team.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot add a member with the owner role (there must be exactly one owner)
        - If a member with this email already exists, the existing member is returned

        Args:
            **kwargs: Request body parameters

        Raises:
            MembersAddMember400Error: Invalid request or member already exists.
            MembersAddMember401Error: Authentication required
            MembersAddMember403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/members",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            400: MembersAddMember400Error,
            401: MembersAddMember401Error,
            403: MembersAddMember403Error,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalMember, method_errors)

    def remove_member(
        self,
        id: str,
    ) -> None:
        """
        Remove a member from the team.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot remove yourself
        - Cannot remove the only owner

        Args:
            id:

        Raises:
            MembersRemoveMember400Error: Cannot remove the only owner.
            MembersRemoveMember401Error: Authentication required
            MembersRemoveMember403Error: Not permitted - requires owner or billing manager role
            MembersRemoveMember404Error: Member not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/customer-portal/members/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            400: MembersRemoveMember400Error,
            401: MembersRemoveMember401Error,
            403: MembersRemoveMember403Error,
            404: MembersRemoveMember404Error,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    def update_member(
        self,
        id: str,
        **kwargs: typing.Unpack[CustomerPortalMemberUpdate],
    ) -> CustomerPortalMember:
        """
        Update a member's name or role.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot modify your own role (to prevent self-demotion)
        - Customer must have exactly one owner at all times

        Args:
            id:
            **kwargs: Request body parameters

        Raises:
            MembersUpdateMember400Error: Invalid role change.
            MembersUpdateMember401Error: Authentication required
            MembersUpdateMember403Error: Not permitted - requires owner or billing manager role
            MembersUpdateMember404Error: Member not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/customer-portal/members/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            400: MembersUpdateMember400Error,
            401: MembersUpdateMember401Error,
            403: MembersUpdateMember403Error,
            404: MembersUpdateMember404Error,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalMember, method_errors)


class MembersAsync(AsyncServiceBase):
    async def list_members(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> ListResourceCustomerPortalMember:
        """
        List all members of the customer's team.

        Only available to owners and billing managers of team customers.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Raises:
            MembersListMembers401Error: Authentication required
            MembersListMembers403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/members",
            path_params={},
            query_params={
                "page": page,
                "limit": limit,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            401: MembersListMembers401Error,
            403: MembersListMembers403Error,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, ListResourceCustomerPortalMember, method_errors
        )

    async def iter_list_members(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> typing.AsyncGenerator[CustomerPortalMember, None]:
        """
        List all members of the customer's team.

        Only available to owners and billing managers of team customers.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Returns:
            An async generator that yields items of type CustomerPortalMember.

        Raises:
            MembersListMembers401Error: Authentication required
            MembersListMembers403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = await self.list_members(
                page=page,
                limit=limit,
            )
            for item in response.items:
                yield item
            if page >= response.pagination.max_page:
                break
            page += 1

    async def add_member(
        self,
        **kwargs: typing.Unpack[CustomerPortalMemberCreate],
    ) -> CustomerPortalMember:
        """
        Add a new member to the customer's team.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot add a member with the owner role (there must be exactly one owner)
        - If a member with this email already exists, the existing member is returned

        Args:
            **kwargs: Request body parameters

        Raises:
            MembersAddMember400Error: Invalid request or member already exists.
            MembersAddMember401Error: Authentication required
            MembersAddMember403Error: Not permitted - requires owner or billing manager role
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/members",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: MembersAddMember400Error,
            401: MembersAddMember401Error,
            403: MembersAddMember403Error,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalMember, method_errors)

    async def remove_member(
        self,
        id: str,
    ) -> None:
        """
        Remove a member from the team.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot remove yourself
        - Cannot remove the only owner

        Args:
            id:

        Raises:
            MembersRemoveMember400Error: Cannot remove the only owner.
            MembersRemoveMember401Error: Authentication required
            MembersRemoveMember403Error: Not permitted - requires owner or billing manager role
            MembersRemoveMember404Error: Member not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/customer-portal/members/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: MembersRemoveMember400Error,
            401: MembersRemoveMember401Error,
            403: MembersRemoveMember403Error,
            404: MembersRemoveMember404Error,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    async def update_member(
        self,
        id: str,
        **kwargs: typing.Unpack[CustomerPortalMemberUpdate],
    ) -> CustomerPortalMember:
        """
        Update a member's name or role.

        Only available to owners and billing managers of team customers.

        Rules:
        - Cannot modify your own role (to prevent self-demotion)
        - Customer must have exactly one owner at all times

        Args:
            id:
            **kwargs: Request body parameters

        Raises:
            MembersUpdateMember400Error: Invalid role change.
            MembersUpdateMember401Error: Authentication required
            MembersUpdateMember403Error: Not permitted - requires owner or billing manager role
            MembersUpdateMember404Error: Member not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/customer-portal/members/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: MembersUpdateMember400Error,
            401: MembersUpdateMember401Error,
            403: MembersUpdateMember403Error,
            404: MembersUpdateMember404Error,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalMember, method_errors)

File: sdk/typescript/src/2026-04/services/customers/members.ts
import type { ClientBase, RequestOptions } from "../../../base";
import type {
  ListResourceMember,
  Member,
  MemberCreateFromCustomer,
  MemberRole,
  MemberSortProperty,
  MemberUpdate,
} from "../../models";

import {
  AmbiguousExternalCustomerID,
  HTTPValidationError,
  NotPermitted,
  ResourceNotFound,
} from "../../errors";

export const listMembers = (client: ClientBase) => {
  /**
   * List the members of a customer.
   *
   * **Scopes**: `members:read` `members:write`
   *
   * @param id - The customer ID.
   * @param query - Query parameters
   * @param requestOptions - Request options
   * @returns {ListResourceMember}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Customer not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    query?: {
      role?: MemberRole | null;
      page?: number;
      limit?: number;
      sorting?: MemberSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): Promise<ListResourceMember> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {
      role: query?.role,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["-created_at"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/customers/{id}/members",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<ListResourceMember>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
/**
 * List the members of a customer.
 *
 * **Scopes**: `members:read` `members:write`
 *
 * @param id - The customer ID.
 * @param query - Query parameters
 * @param requestOptions - Request options
 * @returns {AsyncGenerator<Member>} A generator that yields items of type Member.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {ResourceNotFound} Customer not found.
 * @throws {HTTPValidationError} Validation Error
 */
export const iterListMembers = (client: ClientBase) => {
  return async function* (
    id: string,
    query?: {
      role?: MemberRole | null;
      page?: number;
      limit?: number;
      sorting?: MemberSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): AsyncGenerator<Member> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listMembers(client)(id, { ...query, page, limit }, requestOptions);
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const createMembers = (client: ClientBase) => {
  /**
   * Create a new member for a customer.
   *
   * Only B2B customers with the member management feature enabled can add members.
   * The authenticated user or organization must have access to the customer's organization.
   *
   * **Scopes**: `members:write`
   *
   * @param id - The customer ID.
   * @param body - Request body
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {NotPermitted} Not permitted to add members.
   * @throws {ResourceNotFound} Customer not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    body: MemberCreateFromCustomer,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customers/{id}/members",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      403: NotPermitted,
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const listExternalMembers = (client: ClientBase) => {
  /**
   * List the members of a customer identified by its external ID.
   *
   * **Scopes**: `members:read` `members:write`
   *
   * @param external_id - The customer external ID.
   * @param query - Query parameters
   * @param requestOptions - Request options
   * @returns {ListResourceMember}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Customer not found.
   * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    external_id: string,
    query?: {
      role?: MemberRole | null;
      page?: number;
      limit?: number;
      sorting?: MemberSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): Promise<ListResourceMember> => {
    const pathParams = {
      external_id: external_id,
    };
    const queryParams = {
      role: query?.role,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["-created_at"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/customers/external/{external_id}/members",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<ListResourceMember>(response, "json", {
      404: ResourceNotFound,
      409: AmbiguousExternalCustomerID,
      422: HTTPValidationError,
    });
  };
};
/**
 * List the members of a customer identified by its external ID.
 *
 * **Scopes**: `members:read` `members:write`
 *
 * @param external_id - The customer external ID.
 * @param query - Query parameters
 * @param requestOptions - Request options
 * @returns {AsyncGenerator<Member>} A generator that yields items of type Member.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {ResourceNotFound} Customer not found.
 * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
 * @throws {HTTPValidationError} Validation Error
 */
export const iterListExternalMembers = (client: ClientBase) => {
  return async function* (
    external_id: string,
    query?: {
      role?: MemberRole | null;
      page?: number;
      limit?: number;
      sorting?: MemberSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): AsyncGenerator<Member> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listExternalMembers(client)(
        external_id,
        { ...query, page, limit },
        requestOptions,
      );
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const createExternalMembers = (client: ClientBase) => {
  /**
   * Create a new member for a customer identified by its external ID.
   *
   * **Scopes**: `members:write`
   *
   * @param external_id_path - The customer external ID.
   * @param body - Request body
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {NotPermitted} Not permitted to add members.
   * @throws {ResourceNotFound} Customer not found.
   * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    external_id_path: string,
    body: MemberCreateFromCustomer,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      external_id: external_id_path,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customers/external/{external_id}/members",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      403: NotPermitted,
      404: ResourceNotFound,
      409: AmbiguousExternalCustomerID,
      422: HTTPValidationError,
    });
  };
};
export const getMembers = (client: ClientBase) => {
  /**
   * Get a member of a customer by its ID.
   *
   * **Scopes**: `members:read` `members:write`
   *
   * @param id - The customer ID.
   * @param member_id
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    member_id: string,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      id: id,
      member_id: member_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/customers/{id}/members/{member_id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const deleteMembers = (client: ClientBase) => {
  /**
   * Delete a member of a customer.
   *
   * **Scopes**: `members:write`
   *
   * @param id - The customer ID.
   * @param member_id
   * @param requestOptions - Request options
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string, member_id: string, requestOptions?: RequestOptions): Promise<void> => {
    const pathParams = {
      id: id,
      member_id: member_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "DELETE",
      "/v1/customers/{id}/members/{member_id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<void>(response, "none", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const updateMembers = (client: ClientBase) => {
  /**
   * Update a member of a customer.
   *
   * Only name, email and role can be updated.
   *
   * **Scopes**: `members:write`
   *
   * @param id - The customer ID.
   * @param member_id
   * @param body - Request body
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    member_id: string,
    body: MemberUpdate,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      id: id,
      member_id: member_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/customers/{id}/members/{member_id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const getExternalMembers = (client: ClientBase) => {
  /**
   * Get a member by external ID for a customer identified by its external ID.
   *
   * **Scopes**: `members:read` `members:write`
   *
   * @param external_id - The customer external ID.
   * @param member_external_id - The member external ID.
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    external_id: string,
    member_external_id: string,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      external_id: external_id,
      member_external_id: member_external_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/customers/external/{external_id}/members/{member_external_id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      404: ResourceNotFound,
      409: AmbiguousExternalCustomerID,
      422: HTTPValidationError,
    });
  };
};
export const deleteExternalMembers = (client: ClientBase) => {
  /**
   * Delete a member by external ID for a customer identified by its external ID.
   *
   * **Scopes**: `members:write`
   *
   * @param external_id - The customer external ID.
   * @param member_external_id - The member external ID.
   * @param requestOptions - Request options
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    external_id: string,
    member_external_id: string,
    requestOptions?: RequestOptions,
  ): Promise<void> => {
    const pathParams = {
      external_id: external_id,
      member_external_id: member_external_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "DELETE",
      "/v1/customers/external/{external_id}/members/{member_external_id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<void>(response, "none", {
      404: ResourceNotFound,
      409: AmbiguousExternalCustomerID,
      422: HTTPValidationError,
    });
  };
};
export const updateExternalMembers = (client: ClientBase) => {
  /**
   * Update a member by external ID for a customer identified by its external ID.
   *
   * **Scopes**: `members:write`
   *
   * @param external_id - The customer external ID.
   * @param member_external_id - The member external ID.
   * @param body - Request body
   * @param requestOptions - Request options
   * @returns {Member}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Member not found.
   * @throws {AmbiguousExternalCustomerID} The external customer ID matches customers in several accessible organizations.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    external_id: string,
    member_external_id: string,
    body: MemberUpdate,
    requestOptions?: RequestOptions,
  ): Promise<Member> => {
    const pathParams = {
      external_id: external_id,
      member_external_id: member_external_id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/customers/external/{external_id}/members/{member_external_id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<Member>(response, "json", {
      404: ResourceNotFound,
      409: AmbiguousExternalCustomerID,
      422: HTTPValidationError,
    });
  };
};

export function createMembersService(client: ClientBase) {
  return {
    list: listMembers(client),
    create: createMembers(client),
    listExternal: listExternalMembers(client),
    createExternal: createExternalMembers(client),
    get: getMembers(client),
    delete: deleteMembers(client),
    update: updateMembers(client),
    getExternal: getExternalMembers(client),
    deleteExternal: deleteExternalMembers(client),
    updateExternal: updateExternalMembers(client),
    iterList: iterListMembers(client),
    iterListExternal: iterListExternalMembers(client),
  };
}

export type Members = ReturnType<typeof createMembersService>;
